From e841d1746aa2cbc11cfa00c203b5159ed7f758e8 Mon Sep 17 00:00:00 2001
From: usoock13 <usoock13@gmail.com>
Date: Tue, 1 Nov 2022 00:09:39 +0900
Subject: [PATCH 2/3] player slide melee attack wad added. And modify player
 animation event (it divided attack-end timing and can-move timing).

---
 .../Player/Player Melee Attack 01.anim        | 27 ++++++++++++-------
 .../Player/Player Melee Attack 02.anim        | 13 ++++++---
 .../Player/Player Melee Attack 03.anim        | 13 ++++++---
 Assets/Scripts/Player/MeleePlayer.cs          | 16 +++++++++++
 Assets/Scripts/Player/Player.cs               |  6 +++--
 5 files changed, 57 insertions(+), 18 deletions(-)

diff --git a/Assets/Animations/Player/Player Melee Attack 01.anim b/Assets/Animations/Player/Player Melee Attack 01.anim
index 3918ded..06cc2b0 100644
--- a/Assets/Animations/Player/Player Melee Attack 01.anim	
+++ b/Assets/Animations/Player/Player Melee Attack 01.anim	
@@ -21,17 +21,17 @@ AnimationClip:
   - curve:
     - time: 0
       value: {fileID: 188797140, guid: a6874a41a76525d4aa786fac9a27fd88, type: 3}
-    - time: 0.06666667
+    - time: 0.083333336
       value: {fileID: 1781471063, guid: a6874a41a76525d4aa786fac9a27fd88, type: 3}
-    - time: 0.13333334
-      value: {fileID: -1963189580, guid: a6874a41a76525d4aa786fac9a27fd88, type: 3}
     - time: 0.16666667
+      value: {fileID: -1963189580, guid: a6874a41a76525d4aa786fac9a27fd88, type: 3}
+    - time: 0.23333333
       value: {fileID: -958823570, guid: a6874a41a76525d4aa786fac9a27fd88, type: 3}
-    - time: 0.2
+    - time: 0.3
       value: {fileID: -660099712, guid: a6874a41a76525d4aa786fac9a27fd88, type: 3}
-    - time: 0.23333333
+    - time: 0.36666667
       value: {fileID: -2001238790, guid: a6874a41a76525d4aa786fac9a27fd88, type: 3}
-    - time: 0.5
+    - time: 0.9
       value: {fileID: -2001238790, guid: a6874a41a76525d4aa786fac9a27fd88, type: 3}
     attribute: m_Sprite
     path: 
@@ -64,7 +64,7 @@ AnimationClip:
     m_AdditiveReferencePoseClip: {fileID: 0}
     m_AdditiveReferencePoseTime: 0
     m_StartTime: 0
-    m_StopTime: 0.51666665
+    m_StopTime: 0.9166666
     m_OrientationOffsetY: 0
     m_Level: 0
     m_CycleOffset: 0
@@ -84,21 +84,28 @@ AnimationClip:
   m_HasGenericRootTransform: 0
   m_HasMotionFloatCurves: 0
   m_Events:
-  - time: 0.13333334
+  - time: 0.16666667
     functionName: AnimationEvent_Attack
     data: 
     objectReferenceParameter: {fileID: 0}
     floatParameter: 0
     intParameter: 0
     messageOptions: 0
-  - time: 0.23333333
+  - time: 0.36666667
     functionName: AnimationEvent_AfterAttack
     data: 
     objectReferenceParameter: {fileID: 0}
     floatParameter: 0
     intParameter: 0
     messageOptions: 0
-  - time: 0.5
+  - time: 0.56666666
+    functionName: AnimationEvent_CanMove
+    data: 
+    objectReferenceParameter: {fileID: 0}
+    floatParameter: 0
+    intParameter: 0
+    messageOptions: 0
+  - time: 0.9
     functionName: AnimationEvent_EndAttack
     data: 
     objectReferenceParameter: {fileID: 0}
diff --git a/Assets/Animations/Player/Player Melee Attack 02.anim b/Assets/Animations/Player/Player Melee Attack 02.anim
index b7c52b0..8952624 100644
--- a/Assets/Animations/Player/Player Melee Attack 02.anim	
+++ b/Assets/Animations/Player/Player Melee Attack 02.anim	
@@ -33,7 +33,7 @@ AnimationClip:
       value: {fileID: 308060912, guid: 0ef8ee71eb4bccc4f9f872381df48323, type: 3}
     - time: 0.33333334
       value: {fileID: 1255106716, guid: 0ef8ee71eb4bccc4f9f872381df48323, type: 3}
-    - time: 0.6
+    - time: 0.8666667
       value: {fileID: 1255106716, guid: 0ef8ee71eb4bccc4f9f872381df48323, type: 3}
     attribute: m_Sprite
     path: 
@@ -67,7 +67,7 @@ AnimationClip:
     m_AdditiveReferencePoseClip: {fileID: 0}
     m_AdditiveReferencePoseTime: 0
     m_StartTime: 0
-    m_StopTime: 0.6166667
+    m_StopTime: 0.8833333
     m_OrientationOffsetY: 0
     m_Level: 0
     m_CycleOffset: 0
@@ -101,7 +101,14 @@ AnimationClip:
     floatParameter: 0
     intParameter: 0
     messageOptions: 0
-  - time: 0.6
+  - time: 0.53333336
+    functionName: AnimationEvent_CanMove
+    data: 
+    objectReferenceParameter: {fileID: 0}
+    floatParameter: 0
+    intParameter: 0
+    messageOptions: 0
+  - time: 0.8666667
     functionName: AnimationEvent_EndAttack
     data: 
     objectReferenceParameter: {fileID: 0}
diff --git a/Assets/Animations/Player/Player Melee Attack 03.anim b/Assets/Animations/Player/Player Melee Attack 03.anim
index da26ae0..e7d52a1 100644
--- a/Assets/Animations/Player/Player Melee Attack 03.anim	
+++ b/Assets/Animations/Player/Player Melee Attack 03.anim	
@@ -35,7 +35,7 @@ AnimationClip:
       value: {fileID: -1550687877, guid: da9526b469a81ad45bc60167dbe37184, type: 3}
     - time: 0.45
       value: {fileID: -229533208, guid: da9526b469a81ad45bc60167dbe37184, type: 3}
-    - time: 0.71666664
+    - time: 0.9166667
       value: {fileID: -229533208, guid: da9526b469a81ad45bc60167dbe37184, type: 3}
     attribute: m_Sprite
     path: 
@@ -70,7 +70,7 @@ AnimationClip:
     m_AdditiveReferencePoseClip: {fileID: 0}
     m_AdditiveReferencePoseTime: 0
     m_StartTime: 0
-    m_StopTime: 0.7333333
+    m_StopTime: 0.93333334
     m_OrientationOffsetY: 0
     m_Level: 0
     m_CycleOffset: 0
@@ -104,7 +104,14 @@ AnimationClip:
     floatParameter: 0
     intParameter: 0
     messageOptions: 0
-  - time: 0.71666664
+  - time: 0.65
+    functionName: AnimationEvent_CanMove
+    data: 
+    objectReferenceParameter: {fileID: 0}
+    floatParameter: 0
+    intParameter: 0
+    messageOptions: 0
+  - time: 0.9166667
     functionName: AnimationEvent_EndAttack
     data: 
     objectReferenceParameter: {fileID: 0}
diff --git a/Assets/Scripts/Player/MeleePlayer.cs b/Assets/Scripts/Player/MeleePlayer.cs
index 1758773..afdd742 100644
--- a/Assets/Scripts/Player/MeleePlayer.cs
+++ b/Assets/Scripts/Player/MeleePlayer.cs
@@ -37,6 +37,7 @@ public class MeleePlayer : Player {
         #region Attack State
         attackState01.OnActive += (prevState) => {
             playerAnimator.SetBool("Melee Attack 01", true);
+            playerRigidbody.velocity = new Vector2(0, playerRigidbody.velocity.y);
         };
         attackState01.OnInactive += (State nextState) => {
             playerAnimator.SetBool("Melee Attack 01", false);
@@ -67,6 +68,7 @@ public class MeleePlayer : Player {
             isAfterAttack = false;
             attackingPreInput = false;
             comboCount = 0;
+            canMove = true;
         }
         #endregion Attack State
 
@@ -106,36 +108,47 @@ public class MeleePlayer : Player {
         comboCount++;
         isAfterAttack = false;
         attackingPreInput = false;
+        canMove = false;
+        Vector2 slideforce = new Vector2(transform.localScale.x + moveDirection.x, 0);
         switch(comboCount) {
             case 1:
                 playerStateMachine.ChangeState(attackState01);
+                slideforce *= 3;
                 break;
             case 2:
                 playerStateMachine.ChangeState(attackState02);
+                slideforce *= 3;
                 break;
             case 3:
                 playerStateMachine.ChangeState(attackState03);
+                slideforce *= 5;
                 break;
         }
+        playerRigidbody.AddForce(slideforce, ForceMode2D.Impulse);
     }
     void AnimationEvent_Attack(int index) {
         Collider2D[] inners = Physics2D.OverlapBoxAll(attackRange[index].position, attackRange[index].localScale, 0, Monster.DEFALUT_MONSTER_LAYER);
         float damage = attackDamage;
         float force = attackForce;
+        // Vector2 slideforce = new Vector2(moveDirection.x, 0);
         switch(index) {
             case 0:
                 damage *= 1f;
                 force *= 2f;
+                // slideforce *= 5;
                 break;
             case 1:
                 damage *= 1.2f;
                 force *= 2f;
+                // slideforce *= 5;
                 break;
             case 2:
                 damage *= 1.5f;
                 force *= 5f;
+                // slideforce *= 12;
                 break;
         }
+        // playerRigidbody.AddForce(slideforce, ForceMode2D.Impulse);
         foreach(Collider2D inner in inners) {
             IDamageable target;
             if(inner.TryGetComponent<IDamageable>(out target)) {
@@ -153,6 +166,9 @@ public class MeleePlayer : Player {
             isAfterAttack = true;
         }
     }
+    void AnimationEvent_CanMove() {
+        canMove = true;
+    }
     void AnimationEvent_EndAttack() {
         playerStateMachine.ChangeState(basicState);
         playerAnimator.SetTrigger("End Attack");
diff --git a/Assets/Scripts/Player/Player.cs b/Assets/Scripts/Player/Player.cs
index 7aef196..953bbfd 100644
--- a/Assets/Scripts/Player/Player.cs
+++ b/Assets/Scripts/Player/Player.cs
@@ -28,6 +28,7 @@ public class Player : LivingEntity, IDamageable {
     protected float moveSpeed = 10f;
     protected float jumpPower = 25f;
     protected Vector2 moveDirection;
+    protected bool canMove = true;
     bool isGrounding = false;
     int maxJumpCount = 2;
     int currentJumpCount = 0;
@@ -169,7 +170,8 @@ public class Player : LivingEntity, IDamageable {
     public void Jump() {
         if(currentJumpCount < maxJumpCount
         && !playerStateMachine.Compare(dodgeState)
-        && !playerStateMachine.Compare(hitState)) {
+        && !playerStateMachine.Compare(hitState)
+        && !playerStateMachine.Compare(ATTACK_STATE_TAG)) {
             playerRigidbody.velocity = new Vector2(playerRigidbody.velocity.x, 0);
             playerRigidbody.AddForce(Vector2.up * jumpPower, ForceMode2D.Impulse);
             playerStateMachine.ChangeState(floatState, true);
@@ -239,7 +241,7 @@ public class Player : LivingEntity, IDamageable {
         || playerStateMachine.Compare(floatState)
         || playerStateMachine.Compare(dodgeState)
         || playerStateMachine.Compare(hitState)
-        || playerStateMachine.Compare(ATTACK_STATE_TAG) && !isAfterAttack
+        || playerStateMachine.Compare(ATTACK_STATE_TAG) && !canMove
         || playerStateMachine.Compare(JUMP_ATTACK_STATE_TAG)) return;
 
         if(moveDirection == Vector2.zero) { // Stop Moving
-- 
2.37.3.windows.1

